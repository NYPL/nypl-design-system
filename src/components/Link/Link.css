@import "../Icons/Icon.css";

@define-mixin link {
  color: var(--ui-link-primary);

  &--no-underline {
    text-decoration: none;
  }

  &:hover {
    color: var(--ui-link-secondary);
  }

  /* // .button links have a blue background */
  &.button:hover {
    color: var(--ui-white);
  }

  &[role="button"] {
    cursor: pointer;
  }

  &--dark-background {
    color: var(--ui-white);

    &:hover {
      color: var(--ui-gray-light);
    }
  }
}

@define-mixin more-link {
  align-items: center;
  color: var(--ui-link-primary);
  display: inline-flex;
  justify-content: space-between;
  text-decoration: underline;

  &:hover {
    color: var(--ui-link-secondary);
    text-decoration: underline;
  }

  &__icon {
    @add-mixin icon;

    height: var(--space-s);
    text-decoration: none;
    width: var(--space-s);

    &--left {
      margin-right: var(--space-xs);
    }

    &--right {
      margin-left: var(--space-xs);
    }

    svg {
      height: 100%;
      width: 100%;
    }
  }
}

.nypl-ds {
  a,
  .link {
    @add-mixin link;
  }
}

.more-link {
  @add-mixin more-link;
}
