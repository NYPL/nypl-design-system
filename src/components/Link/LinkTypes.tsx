export enum LinkTypes {
  Action = "action",
  Backwards = "backwards",
  Button = "button",
  Default = "default",
  Forwards = "forwards",
}
