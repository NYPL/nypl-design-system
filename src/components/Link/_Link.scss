@mixin link {
  color: var(--ui-link-primary);

  &--no-underline {
    text-decoration: none;
  }

  &:hover {
    color: var(--ui-link-secondary);
  }

  // .button links have a blue background
  &.button {
    color: var(--ui-white);
  }

  &[role="button"] {
    cursor: pointer;
  }

  &--dark-background {
    color: var(--ui-white);

    &:hover {
      color: var(--ui-gray-light);
    }
  }
}

@mixin more-link {
  align-items: center;
  color: var(--ui-link-primary);
  display: inline-flex;
  justify-content: space-between;
  text-decoration: underline;

  &:hover {
    color: var(--ui-link-secondary);
    text-decoration: underline;
  }

  &__icon {
    @include icon;

    height: var(--space-s);
    text-decoration: none;
    width: var(--space-s);

    &--left {
      margin-right: var(--space-xs);
    }

    &--right {
      margin-left: var(--space-xs);
    }

    svg {
      height: 100%;
      width: 100%;
    }
  }
}

.nypl-ds {
  a,
  .link {
    @include link;
  }
}

.more-link {
  @include more-link;
}
