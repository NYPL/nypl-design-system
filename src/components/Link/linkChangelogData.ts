/** This data is used to populate the ComponentChangelogTable component.
 *
 * date: string (when adding new entry during development, set value as "Prerelease")
 * version: string (when adding new entry during development, set value as "Prerelease")
 * type: "Bug Fix" | "New Feature" | "Update";
 * affects: array["Accessibility" | "Documentation" | "Functionality" | "Styles"];
 * notes: array (will render as a bulleted list, add one array element for each list element)
 */
import { ChangelogData } from "../../utils/ComponentChangelogTable";

export const changelogData: ChangelogData[] = [
  {
    date: "2023-11-14",
    version: "Prerelease",
    type: "Update",
    affects: ["Styles"],
    notes: [
      "Added `hasVisitedStyles` prop which is used to include or omit the component's visited state styles. Default value is true.",
      "Removed `disabled` variant from theme file, as it isn't being used.",
    ],
  },
];
