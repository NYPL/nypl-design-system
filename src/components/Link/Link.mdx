import { Box } from "@chakra-ui/react";
import { ArgTypes, Canvas, Description, Meta, Source } from "@storybook/blocks";
import { cloneElement } from "react";

import Link from "./Link";
import Icon from "../Icons/Icon";
import * as LinkStories from "./Link.stories";

<Meta of={LinkStories} />

# Link

| Component Version | DS Version |
| ----------------- | ---------- |
| Added             | `0.0.4`    |
| Latest            | `1.6.0`    |

## Table of Contents

- {<Link href="#overview" target="_self">Overview</Link>}
- {<Link href="#component-props" target="_self">Component Props</Link>}
- {<Link href="#accessibility" target="_self">Accessibility</Link>}
- {<Link href="#links-with-icons" target="_self">Links With Icons</Link>}
- {<Link href="#anchor-element-rendering" target="_self">Anchor Element Rendering</Link>}
- {<Link href="#link-with-routers" target="_self">Link with Routers</Link>}

## Overview

<Description of={LinkStories} />

## Component Props

<Canvas of={LinkStories.WithControls} />

<ArgTypes of={LinkStories.WithControls} />

## Accessibility

The `Link` component should be used for navigation. If an `onClick` user
interface action is required, a `Button` component should be used instead. The
`Link` component renders an `<a>` element with the `href` attribute.

The `"external"` variant of the `Link` component includes `"screen reader only"`
descriptive text to clarify that the link will open in a new tab/window.
Additionally it renders with the `rel` attribute containing
the relation names "nofollow", "noopener" and "noreferrer".

<Canvas of={LinkStories.Accessibility} />

Resources:

- [W3C WAI Link Examples](https://www.w3.org/TR/wai-aria-practices-1.1/examples/link/link.html)
- [Yale Usability & Web Accessibility Links](https://usability.yale.edu/web-accessibility/articles/links)
- [WebAIM: Invisible Content Just for Screen Reader Users](https://webaim.org/techniques/css/invisiblecontent/)
- [IANA: Link Relation Types](https://www.iana.org/assignments/link-relations/link-relations.xhtml)

## Links With Icons

To render an `Icon` as part of the link, make sure that the `Link` component
wraps the `Icon`. Use `type="action"` to apply appropriate styling to
links with icons. Icons can be rendered to the left or right of the link text.

<Canvas of={LinkStories.LinksWithIcons} />

## Anchor Element Rendering

`Link` can wrap an existing `<a>` element (and `Icon` component) or it
can use the `href` prop to generate an `<a>` element.

<Source
  code={`
// Existing anchor element
<Link type="action">
  <a href="#existing-anchor-tag">link</a>
</Link>
// ...
// Also works with an icon component. Make sure to wrap the icon and anchor
// in a single element for this pattern.
<Link type="action">
  <>
    <Icon name="check" align="left" size="small" />
    <a href="#existing-anchor-tag">check link</a>
  </>
</Link>
// ...
// Otherwise, the \`href\` prop will generate an \`<a>\` tag.
<Link type="action" href="#passed-in-link">
  link
</Link>
`}
  language="jsx"
/>

{

<>
  This is a{" "}
  <Link type="action">
    <a href="#existing-anchor-tag">link</a>
  </Link>{" "}
  with the &lt;a&gt; element as a child of the `Link` component. This is a{" "}
  <Link type="action" href="#passed-in-link">
    link
  </Link>{" "}
  where the `Link` component uses the `href` prop and has a string-only child.
  Finally, this is a{" "}
  <Link type="action">
    <>
      <Icon name="check" align="left" size="small" />
      <a href="#existing-anchor-tag">link</a>
    </>
  </Link>{" "}
  with a check icon.
  <Box mb="20px"></Box>
</>
}

## Link with Routers

### React Router

The Design System's `Link` component should wrap around `react-router`'s own
`Link` component. To avoid name conflicts, rename either the Design System's
or `react-router`'s `Link` component. Any of the following patterns are valid.

<Source
  code={`
// In this first example, React Router's \`Link\` component
// is renamed as \`ReactRouterLink\`.
import {
  BrowserRouter as Router,
  Link as ReactRouterLink,
} from "react-router-dom";
import { Link } from "@nypl/design-system-react-components";
// ...
<Link type="action">
  <ReactRouterLink to="#">Next Page</ReactRouterLink>
</Link>;
`}
  language="jsx"
/>

<Source
  code={`
// In this second example, React Router's \`Link\` component
// is not renamed but the DS's \`Link\` component is renamed
// to \`DSLink\`.
import { BrowserRouter, Link } from "react-router-dom";
import { Link as DSLink } from "@nypl/design-system-react-components";
// ...
<DSLink type="action">
  <Link to="#">Next Page</Link>
</DSLink>;
`}
  language="jsx"
/>

<Source
  code={`
// In this third example, React Router's \`Link\` component
// is not renamed and the DS's \`Link\` component is
// imported and rendered as \`DS.Link\`.
import { BrowserRouter, Link } from "react-router-dom";
import DS from "@nypl/design-system-react-components";
// ...
<DS.Link type="action">
  <Link to="#">Next Page</Link>
</DS.Link>;
`}
  language="jsx"
/>

### NextJS

NextJS's `Link` component should wrap the Design System's `Link` component.
The `href` and `passHref` props MUST be passed into NextJS' `Link` component.
These props, along with an implicit `ref` from NextJS and the Design System's
`Link` component `forwardRef` implementation, handle correctly routing between
pages in a NextJS app.

<Source
  code={`
import { Link } from "@nypl/design-system-react-components";
// ...
/**
 * This is just an example wrapper that works similarly to NextJS' \`Link\`
 * component. In real life, NextJS's \`Link\` component will pass down the
 * \`href\` and \`passHref\` props and a \`ref\` to make routing functional.
 */
export const NextJsLink = (props: HTMLAnchorElement) =>
  React.cloneElement(
    props.children,
    { ...props },
    props.children.props.children
  );
// ...
// Note that NextJS' \`href\` and \`passHref\` props are required.
<NextJsLink href="#" passHref>
  <Link type="action">Next Page</Link>
</NextJsLink>;
`}
  language="jsx"
/>

export const NextJsLink = (props) =>
  cloneElement(
    props.children,
    { href: props.href },
    props.children.props.children
  );

{

<NextJsLink href="#" passHref>
  <Link type="action">Next Page</Link>
</NextJsLink>
}
