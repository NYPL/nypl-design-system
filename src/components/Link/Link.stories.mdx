import {
  Meta,
  Story,
  Canvas,
  ArgsTable,
  Preview,
  Description,
} from "@storybook/addon-docs/blocks";
import { withDesign } from "storybook-addon-designs";
import Link from "./Link";
import {
  BrowserRouter as Router,
  Link as ReactRouterLink,
} from "react-router-dom";
import { LinkTypes } from "./LinkTypes";
import Icon from "../Icons/Icon";

<Meta
  title="Link"
  component={Link}
  decorators={[withDesign]}
  parameters={{
    design: {
      type: "figma",
      url:
        "https://www.figma.com/file/qShodlfNCJHb8n03IFyApM/Master?node-id=16115%3A407",
    },
  }}
  argTypes={{
    blockName: { table: { disable: true } },
    modifiers: { table: { disable: true } },
    children: { table: { disable: true } },
    type: {
      control: {
        type: "select",
      },
    },
  }}
/>

# Link

<Description of={Link} />

<Preview withToolbar>
  <Story
    name="Basic"
    args={{
      id: "nypl-link",
      type: LinkTypes.Action,
      href: "https://nypl.org",
      className: "custom-class",
      attributes: {
        rel: "nofollow",
        onClick: e => e.preventDefault(),
      },
    }}
  >
    {args => <Link {...args}>Link</Link>}
  </Story>
</Preview>

<ArgsTable story="Basic" />

## Links With Icons

To render an `Icon` as part of the link, make sure that the `Link` component
wraps the `Icon`. Use `type={LinkTypes.Action}` to apply appropriate styling to
links with icons. Icons can be rendered to the left or right of the link text.

<Preview>
  <Story name="With Icons">
    <>
      <Link type={LinkTypes.Action} href="#passed-in-link">
        <Icon
          name={"headset"}
          modifiers={["left", "small"]}
          decorative={true}
        />
        Headset Link
      </Link>
      <br />
      <Link type={LinkTypes.Action} href="#passed-in-link">
        <Icon name={"clock"} modifiers={["left", "small"]} decorative={true} />
        Clock Link
      </Link>
      <br />
      <Link type={LinkTypes.Action} href="#passed-in-link">
        <Icon name={"check"} modifiers={["left", "small"]} decorative={true} />
        Check Link
      </Link>
      <br />
      <Link type={LinkTypes.Action} href="#passed-in-link-right">
        Check Link Right
        <Icon name={"check"} modifiers={["right", "small"]} decorative={true} />
      </Link>
    </>
  </Story>
</Preview>

## Anchor Element Rendering

`Link` can wrap an existing `<a>` tag or it can use the `href` prop to generate
an `<a>` tag.

<Preview>
  <Story name="Anchor Element Rendering">
    <>
      This is a{" "}
      <Link type={LinkTypes.Action}>
        <a href="#existing-anchor-tag">link</a>
      </Link>{" "}
      with the &lt;a&gt; tag as a child of the Link component. And this is a{" "}
      <Link type={LinkTypes.Action} href="#passed-in-link">
        link
      </Link>{" "}
      where the Link component uses the `href` prop and has a string-only child.
    </>
  </Story>
</Preview>

## Link with Routers

<Preview>
  <Story name="Link with Routers">
    <div>With React Router</div>
    <Router>
      <Link type={LinkTypes.Action}>
        <ReactRouterLink to="#">Next Page</ReactRouterLink>
      </Link>
    </Router>
    <br />
    <br />
    <div>With NextJS</div>
    <NextJsLink href="#" passHref>
      <Link type={LinkTypes.Action}>Next Page</Link>
    </NextJsLink>
  </Story>
</Preview>

### React Router

The Design System's `Link` component should wrap around `react-router`'s own
`Link` component. To avoid name conflicts, rename either the Design System's
or `react-router`'s `Link` component.

```js
import {
  BrowserRouter as Router,
  Link as ReactRouterLink,
} from "react-router-dom";
import { Link, LinkTypes } from "@nypl/design-system-react-components";

<Link type={LinkTypes.Action}>
  <ReactRouterLink to="#">Next Page</ReactRouterLink>
</Link>;

// or

import { BrowserRouter, Link } from "react-router-dom";
import {
  Link as DSLink,
  LinkTypes,
} from "@nypl/design-system-react-components";

<DSLink type={LinkTypes.Action}>
  <Link to="#">Next Page</Link>
</DSLink>;

// or

import { BrowserRouter, Link } from "react-router-dom";
import DS from "@nypl/design-system-react-components";

<DS.Link type={DS.LinkTypes.Action}>
  <Link to="#">Next Page</Link>
</DS.Link>;
```

<Preview>
  <Router>
    <Link type={LinkTypes.Action}>
      <ReactRouterLink to="#">Next Page</ReactRouterLink>
    </Link>
  </Router>
</Preview>

### NextJS

NextJS's `Link` component should wrap the Design System's `Link` component.
The `href` and `passHref` props MUST be passed into NextJS' `Link` component.
These props, along with an implicit `ref` from NextJS and the Design System's
`Link` component `forwardRef` implementation, handle correctly routing between
pages in a NextJS app.

```js
import { Link, LinkTypes } from "@nypl/design-system-react-components";

/* This is just an example wrapper that works similarly to NextJS' `Link`
 * component. In real life, NextJS's `Link` component will pass down the
 * `href` and `passHref` props and a `ref` to make routing functional.
 */
export const NextJsLink = (props: HTMLAnchorElement) => (
  <div>
    {React.cloneElement(
      props.children,
      { ...props },
      props.children.props.children
    )}
  </div>
);

<NextJsLink href="#" passHref>
  <Link type={LinkTypes.Action}>Next Page</Link>
</NextJsLink>;
```

export const NextJsLink = props => (
  <div>
    {React.cloneElement(
      props.children,
      { ...props },
      props.children.props.children
    )}
  </div>
);

<Preview>
  <NextJsLink href="#" passHref>
    <Link type={LinkTypes.Action}>Next Page</Link>
  </NextJsLink>
</Preview>
