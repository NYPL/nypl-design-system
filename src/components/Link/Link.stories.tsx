import * as React from "react";

import Link from "./Link";
import { LinkTypes } from "./LinkTypes";
import Icon from "../Icons/Icon";
import { IconRotationTypes } from "../Icons/IconTypes";

export default {
    title: "Link",
    component: Link,
};

export const passedInLink = () => (
    <Link linkType={LinkTypes.Default}>
        <a href="#passed-in-link">I'm cold</a>
    </Link>
);

export const generatedLink = () => (
    <Link href="#passed-in-link" linkType={LinkTypes.Default}>
        I'm very cold
    </Link>
);

export const buttonLink = () => (
    <Link linkType={LinkTypes.Button} href="#passed-in-link">
        I'm very cold
    </Link>
);

export const forwardsLink = () => (
    <Link linkType={LinkTypes.Forwards}>
        <a href="#passedinlink">content</a>
    </Link>
);

export const backwardsLink = () => (
    <Link href="#passed-in-link" linkType={LinkTypes.Backwards}>
        content
    </Link>
);

export const actionLinkWithDownloadIcon = () => (
    // To Pass in an icon and its link, make sure that the link tag wraps the icon.
    // The passed-in content must be properly styled.
    <Link linkType={LinkTypes.Action}>
        <a href="#passed-link" style={{ display: "flex" }}>
            <Icon
                name="download"
                blockName="more-link"
                modifiers={["left"]}
                decorative={true}
                iconRotation={IconRotationTypes.rotate0}
            ></Icon>
            Download
        </a>
    </Link>
);
