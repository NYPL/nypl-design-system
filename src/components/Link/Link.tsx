import * as React from "react";
import bem from "../../utils/bem";
import { LinkTypes } from "./LinkTypes";
import Icon, { IconProps } from "../Icons/Icon";
import { IconRotationTypes, IconNames } from "../Icons/IconTypes";

export interface LinkProps {
  /** Additional attributes, such as `rel=nofollow`, to pass to the `<a>` tag. */
  attributes?: { [key: string]: any };
  /** BlockName for use with BEM. See how to work with blockNames and BEM here: http://getbem.com/introduction/ */
  blockName?: string;
  /** className that appears in addition to "link" */
  className?: string;
  /** Href attribute */
  href?: string;
  /** ID */
  id?: string;
  /** Controls the link visualsâ€”action, button, or default. */
  type?: LinkTypes;
  /** Modifiers array for use with BEM. See how to work with modifiers and BEM here: http://getbem.com/introduction/ */
  modifiers?: string[];
  /** Any child node passed to the component. */
  children: React.ReactNode;
}

/**
 * A component that uses an `href` prop or a child anchor element, to create an
 * anchor element with added styling and conventions.
 */
const Link = React.forwardRef<HTMLAnchorElement, LinkProps>((props, ref) => {
  const {
    attributes,
    blockName,
    className,
    href,
    id,
    type = LinkTypes.Default,
    modifiers = [],
    children,
  } = props;

  // Merge the necessary props alongside any extra props for the
  // anchor element.
  const linkProps = {
    id,
    href,
    ...attributes,
  };
  let baseClass = "link";

  let childProps = {};

  if (typeof children === "string" && !href) {
    throw new Error("Link needs prop 'href'");
  }

  if (
    type === LinkTypes.Action ||
    type === LinkTypes.Forwards ||
    type === LinkTypes.Backwards
  ) {
    baseClass = "more-link";
  } else if (type === LinkTypes.Button) {
    baseClass = "button";
  }

  let iconRotation, iconPosition, iconLeft, iconRight;
  // An icon needs a position in order for it to be created and
  // rendered in the link.
  if (type === LinkTypes.Backwards) {
    iconRotation = IconRotationTypes.rotate90;
    iconPosition = "left";
  } else if (type === LinkTypes.Forwards) {
    iconRotation = IconRotationTypes.rotate270;
    iconPosition = "right";
  }

  const navigationIconProps: IconProps = {
    name: IconNames.arrow,
    modifiers: [iconPosition, iconRotation, ...modifiers],
    blockName: "more-link",
    decorative: true,
  };

  if (type === LinkTypes.Backwards) {
    iconLeft = <Icon {...navigationIconProps} />;
  } else if (type === LinkTypes.Forwards) {
    iconRight = <Icon {...navigationIconProps} />;
  }
  const linkClassName = bem(baseClass, modifiers, blockName, [className]);

  if (!props.href) {
    // React Types error makes this fail:  https://github.com/DefinitelyTyped/DefinitelyTyped/issues/32832
    // let children = React.Children.only(props.children);
    if (React.Children.count(props.children) > 1) {
      throw new Error("Please pass only one child into Link");
    }
    const children = props.children[0] ? props.children[0] : props.children;
    childProps = children.props;

    return React.cloneElement(
      children,
      {
        className: linkClassName,
        ...linkProps,
        ...childProps,
        ref,
      },
      [children.props.children]
    );
  } else {
    return React.createElement(
      "a",
      {
        ...linkProps,
        className: linkClassName,
        ref,
      },
      iconLeft,
      props.children,
      iconRight
    );
  }
});

Link.displayName = "Link";

export default Link;
