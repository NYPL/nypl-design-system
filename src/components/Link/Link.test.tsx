import { expect } from "chai";
import * as Enzyme from "enzyme";
import * as React from "react";

import Link from "./Link";
import { LinkTypes } from "./LinkTypes";
import {
  BrowserRouter as Router,
  Link as ReactRouterLink,
} from "react-router-dom";
import Icon from "../Icons/Icon";
import { IconRotationTypes, IconNames } from "../Icons/IconTypes";

describe("Link", () => {
  let wrapper: Enzyme.ReactWrapper<any, any>;

  it("Can pass in icon, text as child and url as props", () => {
    wrapper = Enzyme.mount(
      <Link href="#passed-in-link" type={LinkTypes.Action}>
        <Icon
          name={IconNames.download}
          blockName="more-link"
          modifiers={["left"]}
          decorative={true}
          iconRotation={IconRotationTypes.rotate0}
        ></Icon>
        Download
      </Link>
    );
    expect(wrapper.find("a")).to.have.lengthOf(1);
    expect(wrapper.find("Icon")).to.have.lengthOf(1);
  });

  it("Can pass a link with <a> tag", () => {
    wrapper = Enzyme.mount(
      <Link>
        <a href="#test">Test</a>
      </Link>
    );
    expect(wrapper.find("a")).to.have.lengthOf(1);
  });

  it("Can pass a link with <a> tag and an icon", () => {
    wrapper = Enzyme.mount(
      <Link type={LinkTypes.Action}>
        <a href="#test2">
          <Icon
            name={IconNames.download}
            blockName="more-link"
            modifiers={["left"]}
            decorative={true}
            iconRotation={IconRotationTypes.rotate0}
          ></Icon>
          Test
        </a>
      </Link>
    );
    expect(wrapper.find("a")).to.have.lengthOf(1);
    expect(wrapper.find("Icon")).to.have.lengthOf(1);
  });

  it("Generated back link has icon", () => {
    wrapper = Enzyme.mount(
      <Link href="#passed-in-link" type={LinkTypes.Backwards}>
        content
      </Link>
    );
    expect(
      wrapper.find(".more-link__icon").hasClass("more-link__icon--left")
    ).to.equal(true);
  });

  it("Generated forwards link has icon", () => {
    wrapper = Enzyme.mount(
      <Link href="#passed-in-link" type={LinkTypes.Forwards}>
        content
      </Link>
    );
    expect(
      wrapper.find(".more-link__icon").hasClass("more-link__icon--right")
    ).to.equal(true);
  });

  it("Can pass in text as child and url as props", () => {
    wrapper = Enzyme.mount(<Link href="#test">Test</Link>);
    expect(wrapper.find("a")).to.have.lengthOf(1);
  });

  it("Can pass in a ReactRouter Link", () => {
    wrapper = Enzyme.mount(
      <Router>
        <Link type={LinkTypes.Action}>
          <ReactRouterLink to="#">
            <Icon
              name={IconNames.download}
              blockName="more-link"
              modifiers={["left"]}
              decorative={true}
              iconRotation={IconRotationTypes.rotate0}
            ></Icon>
            Download
          </ReactRouterLink>
        </Link>
      </Router>
    );
    expect(wrapper.find("a")).to.have.lengthOf(1);
  });

  it("throws an error if text is passed but no url is passed", () => {
    expect(() => Enzyme.mount(<Link>Test</Link>)).to.throw(
      "Link needs prop 'href'"
    );
  });

  it("throws an error if more than one child is passed", () => {
    expect(() =>
      Enzyme.mount(
        <Link>
          <Icon
            name={IconNames.download}
            blockName="more-link"
            modifiers={["left"]}
            decorative={true}
            iconRotation={IconRotationTypes.rotate0}
          ></Icon>
          <a href="#test">Test</a>
        </Link>
      )
    ).to.throw("Please pass only one child into Link");
  });

  it("Passes the ref to the input element", () => {
    const ref = React.createRef<HTMLAnchorElement>();
    const container = Enzyme.mount(
      <Link href="/some-link" ref={ref}>
        Go to page
      </Link>
    );
    expect(container.find("a").instance()).to.equal(ref.current);
  });
});
