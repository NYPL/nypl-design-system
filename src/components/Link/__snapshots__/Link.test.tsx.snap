// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link renders the UI snapshot correctly 1`] = `
<a
  className="css-1xdhyk6"
  href="#passed-in-link"
  id="standard-link"
  rel={null}
  target={null}
>
  Standard
</a>
`;

exports[`Link renders the UI snapshot correctly 2`] = `
<a
  className="css-1xdhyk6"
  href="#passed-in-link"
  id="forwards-link"
  rel={null}
  target={null}
>
  Forwards
  <svg
    aria-hidden={true}
    className="chakra-icon more-link css-1grhd2q"
    focusable={false}
    id="forwards-link-icon"
    role="img"
    title="arrow icon"
    viewBox="0 0 24 24"
  >
    <g
      stroke="currentColor"
      strokeWidth="1.5"
    >
      <path
        d="M9,9a3,3,0,1,1,4,2.829,1.5,1.5,0,0,0-1,1.415V14.25"
        fill="none"
        strokeLinecap="round"
      />
      <path
        d="M12,17.25a.375.375,0,1,0,.375.375A.375.375,0,0,0,12,17.25h0"
        fill="currentColor"
        strokeLinecap="round"
      />
      <circle
        cx="12"
        cy="12"
        fill="none"
        r="11.25"
        strokeMiterlimit="10"
      />
    </g>
  </svg>
</a>
`;

exports[`Link renders the UI snapshot correctly 3`] = `
<a
  className="css-1xdhyk6"
  href="#passed-in-link"
  id="backwards-link"
  rel={null}
  target={null}
>
  <svg
    aria-hidden={true}
    className="chakra-icon more-link css-1grhd2q"
    focusable={false}
    id="backwards-link-icon"
    role="img"
    title="arrow icon"
    viewBox="0 0 24 24"
  >
    <g
      stroke="currentColor"
      strokeWidth="1.5"
    >
      <path
        d="M9,9a3,3,0,1,1,4,2.829,1.5,1.5,0,0,0-1,1.415V14.25"
        fill="none"
        strokeLinecap="round"
      />
      <path
        d="M12,17.25a.375.375,0,1,0,.375.375A.375.375,0,0,0,12,17.25h0"
        fill="currentColor"
        strokeLinecap="round"
      />
      <circle
        cx="12"
        cy="12"
        fill="none"
        r="11.25"
        strokeMiterlimit="10"
      />
    </g>
  </svg>
  Backwards
</a>
`;

exports[`Link renders the UI snapshot correctly 4`] = `
<a
  className="css-1xdhyk6"
  href="#passed-in-link"
  id="external-link"
  rel="nofollow"
  target="_blank"
>
  External
  <svg
    aria-hidden={true}
    className="chakra-icon more-link css-1grhd2q"
    focusable={false}
    id="external-link-icon"
    role="img"
    title="action_launch icon"
    viewBox="0 0 24 24"
  >
    <g
      stroke="currentColor"
      strokeWidth="1.5"
    >
      <path
        d="M9,9a3,3,0,1,1,4,2.829,1.5,1.5,0,0,0-1,1.415V14.25"
        fill="none"
        strokeLinecap="round"
      />
      <path
        d="M12,17.25a.375.375,0,1,0,.375.375A.375.375,0,0,0,12,17.25h0"
        fill="currentColor"
        strokeLinecap="round"
      />
      <circle
        cx="12"
        cy="12"
        fill="none"
        r="11.25"
        strokeMiterlimit="10"
      />
    </g>
  </svg>
</a>
`;

exports[`Link renders the UI snapshot correctly 5`] = `
<a
  className="css-1xdhyk6"
  href="#passed-in-link"
  id="button-link"
  rel={null}
  target={null}
>
  Button
</a>
`;

exports[`Link renders the UI snapshot correctly 6`] = `
<a
  className="css-1xdhyk6"
  href="#passed-in-link"
  id="icon-link"
  rel={null}
  target={null}
>
  <svg
    aria-hidden={true}
    className="chakra-icon more-link css-1grhd2q"
    focusable={false}
    id="link-icon"
    role="img"
    title="download icon"
    viewBox="0 0 24 24"
  >
    <g
      stroke="currentColor"
      strokeWidth="1.5"
    >
      <path
        d="M9,9a3,3,0,1,1,4,2.829,1.5,1.5,0,0,0-1,1.415V14.25"
        fill="none"
        strokeLinecap="round"
      />
      <path
        d="M12,17.25a.375.375,0,1,0,.375.375A.375.375,0,0,0,12,17.25h0"
        fill="currentColor"
        strokeLinecap="round"
      />
      <circle
        cx="12"
        cy="12"
        fill="none"
        r="11.25"
        strokeMiterlimit="10"
      />
    </g>
  </svg>
  Download
</a>
`;

exports[`Link renders the UI snapshot correctly 7`] = `
<span
  className="css-0"
>
  <a
    className="css-0"
    href="#existing-anchor-tag"
    id="anchor-link"
    rel={null}
    style={Object {}}
    target={null}
  >
    check link
  </a>
</span>
`;

exports[`Link renders the UI snapshot correctly 8`] = `
<span
  className="css-0"
>
  <svg
    aria-hidden={true}
    className="chakra-icon css-1grhd2q"
    focusable={false}
    id="link-icon"
    role="img"
    title="check icon"
    viewBox="0 0 24 24"
  >
    <g
      stroke="currentColor"
      strokeWidth="1.5"
    >
      <path
        d="M9,9a3,3,0,1,1,4,2.829,1.5,1.5,0,0,0-1,1.415V14.25"
        fill="none"
        strokeLinecap="round"
      />
      <path
        d="M12,17.25a.375.375,0,1,0,.375.375A.375.375,0,0,0,12,17.25h0"
        fill="currentColor"
        strokeLinecap="round"
      />
      <circle
        cx="12"
        cy="12"
        fill="none"
        r="11.25"
        strokeMiterlimit="10"
      />
    </g>
  </svg>
  <a
    href="#existing-anchor-tag"
  >
    check link
  </a>
</span>
`;
